In [1]:
#importing library
import pandas as pd
import numpy as np

In [2]:
import matplotlib.pyplot as plt
import seaborn as sns

In [3]:
#reading data
df=pd.read_csv('train_LZdllcl.csv')
df.head()

employee_id         department     region         education gender  \
0        65438  Sales & Marketing   region_7  Master's & above      f   
1        65141         Operations  region_22        Bachelor's      m   
2         7513  Sales & Marketing  region_19        Bachelor's      m   
3         2542  Sales & Marketing  region_23        Bachelor's      m   
4        48945         Technology  region_26        Bachelor's      m   

  recruitment_channel  no_of_trainings  age  previous_year_rating  \
0            sourcing                1   35                   5.0   
1               other                1   30                   5.0   
2            sourcing                1   34                   3.0   
3               other                2   39                   1.0   
4               other                1   45                   3.0   

   length_of_service  KPIs_met _greaterthan_80percent  awards_won  \
0                  8                                1           0   
1                  4                                0           0   
2                  7                                0           0   
3                 10                                0           0   
4                  2                                0           0   

   avg_training_score  is_promoted  
0                  49            0  
1                  60            0  
2                  50            0  
3                  50            0  
4                  73            0

In [4]:
#checking the proportion of response variable (clear case of imbalacing)
df.is_promoted.value_counts(normalize=True)*100

0    91.482995
1     8.517005
Name: is_promoted, dtype: float64

In [5]:
#count of imbalanced classed
df.is_promoted.value_counts()

0    50140
1     4668
Name: is_promoted, dtype: int64

In [6]:
#columns in the dataframe
df.columns

Index(['employee_id', 'department', 'region', 'education', 'gender',
       'recruitment_channel', 'no_of_trainings', 'age', 'previous_year_rating',
       'length_of_service', 'KPIs_met _greaterthan_80percent', 'awards_won',
       'avg_training_score', 'is_promoted'],
      dtype='object')

In [7]:
#checking the shape of the variable
df.shape

(54808, 14)

In [8]:
#checking the missing values present in the data. 1. eduction , 2. previous_year_rating have missing values
df.isnull().sum()

employee_id                           0
department                            0
region                                0
education                          2409
gender                                0
recruitment_channel                   0
no_of_trainings                       0
age                                   0
previous_year_rating               4124
length_of_service                     0
KPIs_met _greaterthan_80percent       0
awards_won                            0
avg_training_score                    0
is_promoted                           0
dtype: int64

In [9]:
#imputing the missing value education with its mode value which is bachelores
df.education[df.education.isnull()]=df.education.mode()[0]

C:\Users\Vinay\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [10]:
#imputing the missing value previous_year_rating with its mode value which is 3
df.previous_year_rating[df.previous_year_rating.isnull()]=3

C:\Users\Vinay\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [11]:
#creating dummy variables of categorical data
edu=pd.get_dummies(df['education'],drop_first=True)

dept=pd.get_dummies(df['department'],drop_first=True)

reg=pd.get_dummies(df['region'],drop_first=True)

gen=pd.get_dummies(df['gender'],drop_first=True)

rec=pd.get_dummies(df['recruitment_channel'],drop_first=True)

df=pd.concat([df,edu],axis=1)

df=pd.concat([df,dept],axis=1)

df=pd.concat([df,reg],axis=1)

df=pd.concat([df,rec],axis=1)

df=pd.concat([df,gen],axis=1)

In [12]:
#dropping the original columns after creating dummies
df.drop(['education','department','employee_id','gender','region','recruitment_channel'],axis=1,inplace=True)

In [ ]:
#creating dataframes x and y
x=df.drop('is_promoted',axis=1)
y=df['is_promoted']

In [14]:
#reading test data
testoriginal=pd.read_csv('test_2umaH9m.csv')

In [15]:
employee=testoriginal.employee_id

In [16]:
#imputing previous_year_rating in test data
testoriginal.previous_year_rating[testoriginal.previous_year_rating.isnull()]=3

C:\Users\Vinay\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [17]:
#imputing education in test data
testoriginal.education[testoriginal.education.isnull()]=testoriginal.education.mode()[0]

C:\Users\Vinay\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [18]:
# creating test data dummies

edutest=pd.get_dummies(testoriginal['education'],drop_first=True)

depttest=pd.get_dummies(testoriginal['department'],drop_first=True)

regtest=pd.get_dummies(testoriginal['region'],drop_first=True)

gentest=pd.get_dummies(testoriginal['gender'],drop_first=True)

rectest=pd.get_dummies(testoriginal['recruitment_channel'],drop_first=True)

testoriginal=pd.concat([testoriginal,edutest],axis=1)

testoriginal=pd.concat([testoriginal,depttest],axis=1)

testoriginal=pd.concat([testoriginal,regtest],axis=1)

testoriginal=pd.concat([testoriginal,rectest],axis=1)

testoriginal=pd.concat([testoriginal,gentest],axis=1)

In [19]:
#dropping original variable after creating dummies
testoriginal.drop(['education','department','employee_id','gender','region','recruitment_channel'],axis=1,inplace=True)

In [20]:
x_test1=testoriginal
#y_test=testoriginal['is_promoted']


# Decision tree

In [21]:
#importing model selection libary to split train and test
from sklearn.model_selection import train_test_split

In [22]:
#train test split
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2)

In [23]:
#importing tree and confusion matrix
from sklearn import tree
from sklearn.metrics import confusion_matrix

In [24]:
#calling decision tree model and fitting it to train data
clf=tree.DecisionTreeClassifier()
clf.fit(x_train,y_train)

DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=None,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, presort=False, random_state=None,
            splitter='best')

In [25]:
#accuracy of the model on test data
confidance=clf.score(x_test,y_test)

In [26]:
confidance

0.8984674329501916

In [27]:
#predicting on test data
y_pred=clf.predict(x_test)

In [28]:
#pedicted values
y_pred

array([0, 0, 0, ..., 0, 0, 1], dtype=int64)

In [29]:
#creating confusion matrix
c=confusion_matrix(y_test,y_pred)
c

array([[9408,  595],
       [ 518,  441]], dtype=int64)

In [30]:
#Overall Accuracy
(c[0,0]+c[1,1])/np.sum(c)*100

89.84674329501917

In [31]:
#recall
9361/(9361+601)

0.9396707488456133

In [32]:
#Specificity
451/(451+549)

0.451

In [33]:
Precision=c[0,0]/(c[0,0]+[1,0])
Precision

array([0.99989372, 1.        ])

In [34]:
Recall=c[0,0]/(c[0,0]+c[0,1])
Recall

0.940517844646606

In [35]:
Fscore=2*(Recall*Precision)/(Recall+Precision)
Fscore

array([0.96929734, 0.96934728])

In [36]:
#classifivation report
from sklearn.metrics import classification_report
print(classification_report(y_test, y_pred))

             precision    recall  f1-score   support

          0       0.95      0.94      0.94     10003
          1       0.43      0.46      0.44       959

avg / total       0.90      0.90      0.90     10962



#  GBM

In [37]:
#import libraries
from sklearn.ensemble import GradientBoostingClassifier #For Classification
from sklearn.ensemble import GradientBoostingRegressor #For Regression



In [38]:
#use GBM function
clf = GradientBoostingClassifier(n_estimators=100, learning_rate=1.0, max_depth=3)
clf.fit(x_train, y_train)

GradientBoostingClassifier(criterion='friedman_mse', init=None,
              learning_rate=1.0, loss='deviance', max_depth=3,
              max_features=None, max_leaf_nodes=None,
              min_impurity_decrease=0.0, min_impurity_split=None,
              min_samples_leaf=1, min_samples_split=2,
              min_weight_fraction_leaf=0.0, n_estimators=100,
              presort='auto', random_state=None, subsample=1.0, verbose=0,
              warm_start=False)

#### Now predicting values for test data

In [39]:
predictions=clf.predict(x_test)


In [40]:
from sklearn.metrics import classification_report
print(classification_report(y_test, y_pred))

             precision    recall  f1-score   support

          0       0.95      0.94      0.94     10003
          1       0.43      0.46      0.44       959

avg / total       0.90      0.90      0.90     10962



In [41]:
predictions=pd.DataFrame(predictions,columns=['is_promoted'])
employee=pd.DataFrame(employee,columns=['employee_id'])

In [42]:
predictions=pd.concat([employee,predictions],axis=1)

In [43]:
#saving prediction to csv
predictions.to_csv('xplorer.csv',index=False)#,columns=['employee_id','is_promoted'])

# Logistic Regression

In [44]:
#importing linear model
from sklearn.linear_model import LogisticRegression
from sklearn import metrics

In [45]:
#calling logistic regression model
logreg = LogisticRegression()
logreg.fit(x_train, y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)

In [46]:
#pedicting on test data and printing accuracy
y_pred = logreg.predict(x_test)
print('Accuracy of logistic regression classifier on test set: {:.2f}'.format(logreg.score(x_test, y_test)))

Accuracy of logistic regression classifier on test set: 0.93


In [47]:
#classification report
from sklearn.metrics import classification_report
print(classification_report(y_test, y_pred))

             precision    recall  f1-score   support

          0       0.93      1.00      0.96     10003
          1       0.83      0.19      0.31       959

avg / total       0.92      0.93      0.90     10962



## Logistic regression using statsmodel

In [ ]:
# import statsmodels.api as sm
logit_model=sm.Logit(y_train,x_train)
result=logit_model.fit()
print(result.summary2())

## Using SMOTE to balance the data

In [49]:
#X = .loc[:, data_final.columns != 'is_promoted']
#y = data_final.loc[:, data_final.columns == 'is_promoted']
from imblearn.over_sampling import SMOTE
os = SMOTE(random_state=0)
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.3, random_state=0)
columns = x_train.columns
os_data_X,os_data_y=os.fit_sample(x_train, y_train)
os_data_X = pd.DataFrame(data=os_data_X,columns=columns )
os_data_y= pd.DataFrame(data=os_data_y,columns=['y'])
# we can Check the numbers of our data
print("length of oversampled data is ",len(os_data_X))
print("Number of no subscription in oversampled data",len(os_data_y[os_data_y['y']==0]))
print("Number of subscription",len(os_data_y[os_data_y['y']==1]))
print("Proportion of no subscription data in oversampled data is ",len(os_data_y[os_data_y['y']==0])/len(os_data_X))
print("Proportion of subscription data in oversampled data is ",len(os_data_y[os_data_y['y']==1])/len(os_data_X))


length of oversampled data is  70166
Number of no subscription in oversampled data 35083
Number of subscription 35083
Proportion of no subscription data in oversampled data is  0.5
Proportion of subscription data in oversampled data is  0.5


## Using RFE for feature selection

In [50]:
#impoting feature selection librray
from sklearn.feature_selection import RFE
from sklearn.linear_model import LogisticRegression

#using balanced data and applied rfe
logreg = LogisticRegression()
rfe = RFE(logreg, 20)
rfe = rfe.fit(os_data_X, os_data_y.values.ravel())
print(rfe.support_)
print(rfe.ranking_)

[False False False False  True  True False False False False  True  True
 False False False False False False  True False False False False  True
 False  True  True False  True  True False False  True False  True False
 False  True False False  True  True  True  True False  True  True False
 False  True False False False]
[15 32  3 33  1  1 20 23 28 16  1  1 17 19  2 18 12 10  1  6  9  4  8  1
 24  1  1 11  1  1 22 25  1 21  1  7 31  1 34 13  1  1  1  1 14  1  1 26
  5  1 29 30 27]


In [51]:
mask = rfe.get_support() #list of booleans
new_features = [] # The list of your K best features

for bool, feature in zip(mask, x):
    if bool:
        new_features.append(feature)

In [52]:
new_features

['KPIs_met _greaterthan_80percent',
 'awards_won',
 'HR',
 'Legal',
 'region_11',
 'region_16',
 'region_18',
 'region_19',
 'region_20',
 'region_21',
 'region_24',
 'region_26',
 'region_29',
 'region_31',
 'region_32',
 'region_33',
 'region_34',
 'region_5',
 'region_6',
 'region_9']

In [53]:
#creating new dataframe X-Y
X=os_data_X[new_features]
y=os_data_y['y']

In [54]:
#train test split and applying logistic model
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=0)
logreg = LogisticRegression()
logreg.fit(X_train, y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)

In [55]:
#prediction on test data
y_pred = logreg.predict(X_test)
print('Accuracy of logistic regression classifier on test set: {:.2f}'.format(logreg.score(X_test, y_test)))

Accuracy of logistic regression classifier on test set: 0.72


In [56]:
#Confusion matrix
confusion_matrix = confusion_matrix(y_test, y_pred)
print(confusion_matrix)

[[7309 3181]
 [2771 7789]]


In [57]:
#classification repot
from sklearn.metrics import classification_report
print(classification_report(y_test, y_pred))

             precision    recall  f1-score   support

          0       0.73      0.70      0.71     10490
          1       0.71      0.74      0.72     10560

avg / total       0.72      0.72      0.72     21050



## Random Forest with balanced data

In [59]:
#importing ensemble models
import sklearn.ensemble as ensemble
from sklearn.model_selection import train_test_split

In [60]:
randomForest2=ensemble.RandomForestClassifier(n_estimators=50,max_features=0.5,oob_score=True)

In [61]:
randomForest2.fit(X_train,y_train)

RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=None, max_features=0.5, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=50, n_jobs=1,
            oob_score=True, random_state=None, verbose=0, warm_start=False)

In [62]:
randomForest2.base_estimator_

DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=None,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, presort=False, random_state=None,
            splitter='best')

In [63]:
pred4=randomForest2.predict(X_test)
pred4[1:5]

array([1, 1, 1, 0], dtype=int64)

In [64]:
from scipy.stats import itemfreq
itemfreq(pred4)

C:\Users\Vinay\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: DeprecationWarning: `itemfreq` is deprecated!
`itemfreq` is deprecated and will be removed in a future version. Use instead `np.unique(..., return_counts=True)`
  


array([[    0,  9363],
       [    1, 11687]], dtype=int64)

In [65]:
randomForest2.feature_importances_     # highest value gives reduction in gini for classification model|||||| for regression -RSS


array([5.63085140e-01, 1.42265926e-01, 3.52454459e-02, 1.35518939e-02,
       2.77776384e-02, 2.95677656e-02, 1.19398844e-04, 1.81712020e-02,
       1.22728020e-02, 5.29364493e-03, 9.39978514e-03, 4.23824635e-02,
       1.58933673e-02, 3.65716282e-02, 1.48360780e-02, 3.54532945e-03,
       3.75460352e-03, 9.43841034e-03, 1.18662552e-02, 4.96122169e-03])

In [66]:
randomForest2.oob_score_

0.8200179167684665

In [68]:
randomForest2.score(X_test,y_test)    #returns the mean accuracy on the given test data and labels

0.8202850356294537

In [69]:
from sklearn.metrics import classification_report
print(classification_report(y_test, pred4))

             precision    recall  f1-score   support

          0       0.86      0.77      0.81     10490
          1       0.79      0.87      0.83     10560

avg / total       0.82      0.82      0.82     21050



## Randon forest using imbalanced data

In [70]:
m=df.drop('is_promoted',axis=1)
n=df['is_promoted']

In [71]:
m_train, m_test, n_train,n_test = train_test_split(m, n, test_size=0.3, random_state=0)

In [72]:
randomForest3=ensemble.RandomForestClassifier(n_estimators=50,max_features=0.5,oob_score=True)

In [76]:
randomForest3.fit(m_train,n_train)

RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=None, max_features=0.5, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=50, n_jobs=1,
            oob_score=True, random_state=None, verbose=0, warm_start=False)

In [77]:
randomForest3.base_estimator_

DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=None,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, presort=False, random_state=None,
            splitter='best')

In [79]:
pred5=randomForest3.predict(m_test)
pred5[1:5]

array([0, 0, 0, 0], dtype=int64)

In [80]:
from scipy.stats import itemfreq
itemfreq(pred5)

C:\Users\Vinay\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: DeprecationWarning: `itemfreq` is deprecated!
`itemfreq` is deprecated and will be removed in a future version. Use instead `np.unique(..., return_counts=True)`
  


array([[    0, 15812],
       [    1,   631]], dtype=int64)

In [81]:
randomForest3.feature_importances_     # highest value gives reduction in gini for classification model|||||| for regression -RSS


array([0.02367143, 0.12375228, 0.05623037, 0.09310164, 0.04671917,
       0.02599601, 0.27810414, 0.00186294, 0.01946182, 0.00971275,
       0.00648319, 0.00273921, 0.02480873, 0.02408118, 0.00407472,
       0.05234885, 0.01368349, 0.00285849, 0.00368204, 0.00181128,
       0.00642591, 0.00347537, 0.00676486, 0.00415304, 0.00447264,
       0.        , 0.00322362, 0.01243233, 0.00326739, 0.00176428,
       0.00983589, 0.00515157, 0.002165  , 0.00440358, 0.0057275 ,
       0.00506633, 0.00471631, 0.00323509, 0.00224855, 0.00303791,
       0.00515013, 0.00229827, 0.00092026, 0.00040162, 0.00594385,
       0.00200105, 0.00193844, 0.00937285, 0.00241395, 0.0008921 ,
       0.00666342, 0.03026688, 0.02498631])

In [82]:
randomForest3.oob_score_

0.9344715235240454

In [83]:
randomForest3.score(m_test,n_test)    #returns the mean accuracy on the given test data and labels

0.9396095602992155

In [84]:
from sklearn.metrics import classification_report
print(classification_report(n_test, pred5))

             precision    recall  f1-score   support

          0       0.94      0.99      0.97     15057
          1       0.81      0.37      0.51      1386

avg / total       0.93      0.94      0.93     16443



# ---------------------------------------------------------------------------------------------------------------

Since the accuracy is not a good measure when we are dealing with imbalanced data for predicting minority class, so we should go for recall/ precision and f1-score.

--Here, we have taken the imbalanced data with no balancing technique applied to the data.

--Random forest has been built on this data.

--Though the accuracy has been increased from random forest model( with balanced data) as it is predicting majority class correctly, (0.82 in balanced data to 0.93 in imbalanced data)

but recall has been dropped considerably from 0.87 to 0.37 :

- Random forest model with balanced data:
               precision    recall  f1-score   support

          0       0.86      0.77      0.81     10490
          1       0.79      0.87      0.83     10560

        avg / total       0.82      0.82      0.82     21050

- Random forest model with imbalanced data:

               precision    recall  f1-score   support

          0       0.94      0.99      0.97     15057
          1       0.81      0.37      0.51      1386

        avg / total       0.93      0.94      0.93     16443

